### PR TITLE
feat(spool): Check preconditions before unspooling

### PR DESCRIPTION
**Pull request by jjbayer**
In the new spooler design, all envelopes go into the envelope buffer unconditionally (as long as it has capacity), but we must check memory capacity and global config readiness before unspooling:

image

ref: https://github.com/getsentry/team-ingest/issues/345

#skip-changelog

### CODE REVIEW

File: relay-server/src/endpoints/common.rs [hunk 1]
```diff
@@ -353,9 +353,7 @@ pub async fn handle_envelope(
     }
 
     // TODO(jjbayer): Remove this check once spool v1 is removed.
-    if state.memory_checker().check_memory().is_exceeded() {
-        // NOTE: Long-term, we should not reject the envelope here, but spool it to disk instead.
-        // This will be fixed with the new spool implementation.
+    if state.envelope_buffer().is_none() && state.memory_checker().check_memory().is_exceeded() {
```
Comment (jjbayer) [inline relay-server/src/endpoints/common.rs:353]: In V2 mode, this check is harmful and can lead to unnecessary event loss.

File: relay-server/src/services/buffer/mod.rs [hunk 2]
```diff
@@ -127,19 +133,38 @@ impl EnvelopeBufferService {
     }
 
     /// Wait for the configured amount of time and make sure the project cache is ready to receive.
-    async fn ready_to_pop(&mut self) -> Result<(), SendError> {
+    async fn sleep(&mut self) -> Result<(), SendError> {
```
Comment (iambriccardo) [inline relay-server/src/services/buffer/mod.rs]: nit: it's a bit of a misleading name, I would rather call this, `ready_to_pop`.

File: relay-server/src/services/buffer/mod.rs [hunk 3]
```diff
@@ -127,19 +133,38 @@ impl EnvelopeBufferService {
...
+
+        self.global_config_rx.borrow().is_ready()
+    }
+
     /// Tries to pop an envelope for a ready project.
     async fn try_pop(
         &mut self,
         buffer: &mut PolymorphicEnvelopeBuffer,
     ) -> Result<(), EnvelopeBufferError> {
+        if !self.should_pop(buffer) {
```
Comment (iambriccardo) [inline relay-server/src/services/buffer/mod.rs]: Isn't it better if we just return pending in the `sleep` function? To immediately try and handle an incoming message? Or is this done on purpose?

**Review by iambriccardo — APPROVED** [review]
LGTM

File: relay-server/src/services/buffer/mod.rs (hunk 3 above)
Comment (jjbayer) [reply relay-server/src/services/buffer/mod.rs]: Good idea to unite those two functions, the mixture of sync checks and async sleeps makes this logic awkward.

Comment (iambriccardo) [reply relay-server/src/services/buffer/mod.rs]: Exactly, I would rather have a method that does all of that and then you `try_pop`.

**Pull request merged**